### PR TITLE
Windows warnings

### PR DESCRIPTION
**Pull request by huitema**

### CODE REVIEW

File: include/picotls.h [hunk 1]
```diff
@@ -239,7 +239,15 @@ typedef const struct st_ptls_aead_algorithm_t {
     /**
      * the underlying key stream
      */
+#ifdef _WINDOWS
+    /*  suppress warning C4114: same type qualifier used more than once */
+#pragma warning( push )
+#pragma warning (disable : 4114 )
+#endif
     const ptls_cipher_algorithm_t *ctr_cipher;
```
Comment (kazuho) [inline include/picotls.h]: How about removing `const` from this line instead?

File: include/picotls.h [hunk 2]
```diff
@@ -644,9 +661,9 @@ int ptls_buffer_push_asn1_ubigint(ptls_buffer_t *buf, const void *bignum, size_t
         ptls_buffer_push_asn1_block((buf), block);                                                                                 \
     } while (0)
 
-int ptls_decode16(uint16_t *value, const uint8_t **src, const uint8_t *end);
-int ptls_decode32(uint32_t *value, const uint8_t **src, const uint8_t *end);
-int ptls_decode64(uint64_t *value, const uint8_t **src, const uint8_t *end);
+int ptls_decode16(uint16_t *value, const uint8_t **src, const uint8_t *const end);
+int ptls_decode32(uint32_t *value, const uint8_t **src, const uint8_t *const end);
+int ptls_decode64(uint64_t *value, const uint8_t **src, const uint8_t *const end);
```
Comment (kazuho) [inline include/picotls.h]: Would you mind elaborating why this change is necessary?

My understanding is that whether `end` is modified (to clarify: not `*end` but `end`) is an implementation detail and there is no need to express that in the function prototype.

File: include/picotls.h [hunk 3]
```diff
@@ -862,7 +879,15 @@ static ptls_iovec_t ptls_iovec_init(const void *p, size_t len);
...
 {
+#ifdef _WINDOWS
+/* suppress warning C4204: nonstandard extension used: non-constant aggregate initializer */
+#pragma warning( push )
+#pragma warning (disable : 4204 )
+#endif
     return (ptls_iovec_t){(uint8_t *)p, len};
+#ifdef _WINDOWS
+#pragma warning( pop )
+#endif
```
Comment (kazuho) [inline include/picotls.h]: How about rewriting the function body so that we do not need a pragma?

File: include/picotls.h (hunk 2 above)
Comment (huitema) [reply include/picotls.h]: Less necessary for me, as the warning only happens when you compile picotls, not when you just include picotls.h

File: include/picotls.h (hunk 1 above)
Comment (huitema) [reply include/picotls.h]: That would work too. I didn't wang to change your code :-)

File: include/picotls.h (hunk 3 above)
Comment (huitema) [reply include/picotls.h]: Yes. Ifvyou are ok with that I can do it.

File: include/picotls.h (hunk 2 above)
Comment (huitema) [reply include/picotls.h]: The point is, Visual Studio complains when the function's signature does not exactly match the prototype. I understand that this is not true with every compiler. There are three plausible fixes;
1) What I did, change the prototype;
2) Change the signature in the implementation file to match the prototype. Just because a variable is not declared *const does not mean the function should change it.
3) Add "pragma warning(suppress)" in picotls.c above the implementation.
Option 2 might be the simplest.

File: include/picotls.h (hunk 1 above)
Comment (kazuho) [reply include/picotls.h]: I would say that having `const` in this line was a mistake. Let's remove it.

File: include/picotls.h [hunk 4]
```diff
@@ -544,6 +557,10 @@ typedef struct st_ptls_handshake_properties_t {
      */
     int (*collected_extensions)(ptls_t *tls, struct st_ptls_handshake_properties_t *properties, ptls_raw_extension_t *extensions);
 } ptls_handshake_properties_t;
+#ifdef _WINDOWS
+#pragma warning( pop )
+#endif
+
```
Comment (huitema) [inline include/picotls.h]: To eliminate the need for these pragmas, we would have to name the union, but then it would impact a lot of code.

File: include/picotls.h (hunk 2 above)
Comment (kazuho) [reply include/picotls.h]: Let's go with 1. It could be considered a issue of the compiler, but I am comfortable with the change here.

File: include/picotls.h (hunk 3 above)
Comment (kazuho) [reply include/picotls.h]: Please. I am fine with having a few extra lines of code here to avoid `#if`.

File: include/picotls.h (hunk 4 above)
Comment (kazuho) [reply include/picotls.h]: I agree. I think we need this.

**Comment by huitema** [comment]
I agree with the concern of minimizing the ugliness of "pragma" statements. Please look at the modified version.

**Comment by kazuho** [comment]
Thank you for the changes. Merged to master.
